### DISCUSSION TITLE
Using native events wherever one is suitable

### DISCUSSION THREAD

**Discussion by LeaVerou** [shoelace-style/webawesome#193] [Announcements]
Currently, all WA components use their own special events even where a native counterpart exists, `wa-input` instead of `input`, `wa-change` instead of `change` and so on. E.g. here is the list of events that `<wa-input>` emits:

<img width="877" alt="image" src="https://github.com/user-attachments/assets/065e7301-33c5-4d24-b0e0-df8cd7be93e0">

With the exception of `wa-clear`, all of these have native counterparts.

I see several advantages to using the native events:
- Paving the cowpaths: Fewer roadblocks on the path to adopting WA on an existing website. Going from `<input>` to `<wa-input>` requires fewer code changes. 
- Fewer things to learn too, your existing knowledge about how the DOM works can be applied here too.
- In general, as a design goal, good components behave as closely to native elements as possible (with the exception of things where the divergence is a clear improvement, see the TAG [Consistency](https://www.w3.org/TR/design-principles/#consistency) principle)
- Easier to handle form controls generically, without having to care which ones are WA and which ones are native, including via event delegation.
- Libraries with existing heuristics around the DOM still work. E.g. Vue data binding works out of the box with a component that implements an actual `input` event, no plumbing needed.

I don’t remember what the downsides are, but I recall @claviska felt strongly _against_ using native events, so I’m opening this so we can discuss.

**Comment by claviska** [comment]
One disadvantage is how retargeting works. If we rely on native events, we have to make sure EVERY ELEMENT inside EVERY COMPONENT isn't dispatching a conflicting event that might leak out, be it a click, a mouseover, a touchstart, or anything else. Custom events, or more appropriately, _prefixed events_, don't have this risk and can be more clearly defined and fired off without any concern of conflict.

The answer to _should we use native events_ is probably yes at some point, but we first need a way to reliably turn off retargeting for some (or all) events to guarantee we're dispatching only the ones we intend to. It feels like something we should explore for WA4 with an optional migration module that can re-enable the prefixes for backwards compat.

Another issue that prefixing doesn't entirely solve, but somewhat alleviates, is an educational problem. This is probably the number one thing people open issues for:

https://www.abeautifulsite.net/posts/custom-event-names-and-the-bubbling-problem/

And while the same thing happens to native events, it happens a lot less right now because they're not using native events. It's especially painful for React users who don't seem to understand event bubbling in general. 

Another [possible?] issue is collisions with other components that don't prefix, but maybe this isn't important since native elements dispatch them too. This argument probably ends up circling back to the bubbling problem, tbh.

For fun, I made a poll to see if anyone else in my circle has strong opinions on this: https://x.com/cory_laviska/status/1853515918673879322

**Reply by LeaVerou** [reply]
That sounds like we are making the API worse to make our job easier, which we have a design principle against 😁 

Could the retargeting issue be solved by attaching an event listener to the host (which we could do generically via `a handleEvent()` method) and then using [`event.stopImmediatePropagation()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopImmediatePropagation) on any events we don't want to leak? Given that we are the first ones to get a reference to the host, this should always be fired before any author listeners.

WRT bubbling, indeed, as you point out that’s a native issue as well. That is why certain native events do not bubble, e.g. [`toggle`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailsElement/toggle_event#event_type), which is used by both `<details>` and the Popover API. So in your particular example, moving to the native event would _completely_ alleviate that problem, not just reduce its incidence. 😁 

> Another [possible?] issue is collisions with other components that don't prefix, but maybe this isn't important since native elements dispatch them too.

This is only an issue when the events are used incorrectly. E.g. an `input` event should mean the same thing on a native element, a WA component, or a third-party component. If it’s used this way, I don't see it as a "collision"; it's just using the platform rather than inventing new primitives. 

I see it as no different than using global attributes on CEs. It’s not like we use `wa-id` instead of `id`. 

> For fun, I made a poll to see if anyone else in my circle has strong opinions on this:

Right now it seems `input` is actually winning by a bit, though it's still early. I boosted it too. However, if people think using the native events comes with certain baggage like the things mentioned above (e.g. bubbling or leaks), that could bias the results of the poll. And seeing that some people say things like "I know the right way is to namespace" makes me think they have not thought it through very much…

----

Note that this is about using unprefixed event names when there is a suitable native event; not about unprefixing all events. I don't have a strong opinion on whether events with no native equivalent should be prefixed or not, though I will point out we don't prefix element attributes, so why would we prefix its events? OTOH, there could be an actual collision if the web platform later adds an event with that name that behaves incompatibly.

**Reply by claviska** [reply]
I'll admit that it's becoming a best practice in the community to dispatch `input` and `change` events for FACE because it "just works" with Vue and other frameworks.

I'm very open to this, but I'm less certain about non-standard events and what those would be named. For example, we dispatch `sl-show` and `sl-after-show` for `<sl-dialog>`. What would those be called? (Aside: I thought I saw a proposal for a "before show" event for `<dialog>` but I can't find it now — I'd prefer to align with that, but there are other events that don't align to current or future platform events, such as carousel's `sl-slide-change`.)

And, of course, timelines. It's not a difficult change, but it's tedious and these things add up so I want to be cautious of that.

**Reply by LeaVerou** [reply]
Agreed; as I wrote above, it's less clear about non-standard events. Specifically, for `<sl-dialog>` we can use the same events that native `<dialog>` fires. But to your bigger point, I'm fine keeping any nonstandard events namespaced. I'd avoid dashes after the namespace though, for consistency with native events (and to make which part is a namespace more clear), see https://www.w3.org/TR/design-principles/#naming-consistency

> And, of course, timelines. It's not a difficult change, but it's tedious and these things add up so I want to be cautious of that.

Agreed, though if we make it later we may end up having to maintain both…

**Comment by LeaVerou** [comment]
Let's close this, since it's being discussed in #162.
